# Here are some resources to help you get started
[Data Source​ Website](https://data.cms.gov/provider-data/dataset/mj5m-pzi6)

Data Dictionary: https://data.cms.gov/provider-data/sites/default/files/data_dictionaries/physician/DOC_Data_Dictionary.pdf

[Exploratory Data Analysis Tutorials](https://www.analyticsvidhya.com/blog/2021/02/introduction-to-exploratory-data-analysis-eda/)



In [ ]:
import numpy as np
import pandas as pd
import matplotlib as plt
import seaborn as sns
from pylab import rcParams
rcParams['figure.figsize'] = (12,6)
sns.set()

In [ ]:
# This cell downloads the data from the cms website
! wget https://data.cms.gov/provider-data/sites/default/files/resources/69a75aa9d3dc1aed6b881725cf0ddc12_1654518159/DAC_NationalDownloadableFile.csv

--2022-07-05 21:48:27--  https://data.cms.gov/provider-data/sites/default/files/resources/69a75aa9d3dc1aed6b881725cf0ddc12_1654518159/DAC_NationalDownloadableFile.csv
Resolving data.cms.gov (data.cms.gov)... 184.31.25.37, 2600:1406:1400:699::28a, 2600:1406:1400:691::28a
Connecting to data.cms.gov (data.cms.gov)|184.31.25.37|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 735575186 (701M) [text/csv]
Saving to: ‘DAC_NationalDownloadableFile.csv’

DAC_NationalDownloa 100%[===================>] 701.50M  35.4MB/s    in 19s     

2022-07-05 21:48:47 (37.2 MB/s) - ‘DAC_NationalDownloadableFile.csv’ saved [735575186/735575186]



In [ ]:
# Running ls shows us that the DAC_NationalDownnloadableFile.csv is available in the local filesystem for this notebook
! ls

DAC_NationalDownloadableFile.csv  sample_data


In [ ]:
# This line reads in the file to a pandas dataframe
df = pd.read_csv('DAC_NationalDownloadableFile.csv',encoding_errors='ignore',low_memory=False)

In [ ]:
# Now we view the top of the dataframe
df.head()

NPI   Ind_PAC_ID      Ind_enrl_ID    lst_nm  frst_nm  mid_nm  suff  \
0  1215155429   6507031499  I20111205000151    WEYMAN    TERRY       L   NaN   
1  1215377635   9133419120  I20200609000011    DUPONT    BRETT   JAMES   NaN   
2  1215151303   4587852165  I20101227000044    MORGAN  BARBARA       K   NaN   
3  1215148507    749332930  I20120404001029  BRANTLEY    ELISE       I   NaN   
4  1215300983   9638476138  I20160329001230   HUNTLEY   TAYLOR     NaN   NaN   

   gndr  Cred                                       Med_sch  ...  \
0     M   NaN  CLEVELAND CHIROPRACTIC COLLEGE - LOS ANGELES  ...   
1     M   NaN                                         OTHER  ...   
2     F   NaN                                         OTHER  ...   
3     F   NaN    UNIVERSITY OF OKLAHOMA COLLEGE OF MEDICINE  ...   
4     M   NaN                                         OTHER  ...   

    hosp_afl_lbn_2  hosp_afl_3  hosp_afl_lbn_3  hosp_afl_4  hosp_afl_lbn_4  \
0              NaN         NaN             NaN         NaN             NaN   
1              NaN         NaN             NaN         NaN             NaN   
2              NaN         NaN             NaN         NaN             NaN   
3              NaN         NaN             NaN         NaN             NaN   
4              NaN         NaN             NaN         NaN             NaN   

   hosp_afl_5  hosp_afl_lbn_5  ind_assgn   grp_assgn  \
0         NaN             NaN          Y           M   
1         NaN             NaN          Y           M   
2         NaN             NaN          Y           M   
3         NaN             NaN          Y           M   
4         NaN             NaN          Y           M   

                     adrs_id  
0  CA913612415WE2277XRDXX303  
1  CA954767046SO651XXWXXX406  
2  TX756054636LO2401XRDXX301  
3  OK731341849OK3030XSTXX401  
4  SC295014173FL1451XSTXX400  

[5 rows x 40 columns]

In [ ]:
# Here we check how big the dataframe is
df.shape

(2383035, 40)

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2383035 entries, 0 to 2383034
Data columns (total 40 columns):
 #   Column           Dtype  
---  ------           -----  
 0   NPI              int64  
 1    Ind_PAC_ID      int64  
 2    Ind_enrl_ID     object 
 3    lst_nm          object 
 4    frst_nm         object 
 5    mid_nm          object 
 6    suff            object 
 7    gndr            object 
 8    Cred            object 
 9    Med_sch         object 
 10   Grd_yr          float64
 11   pri_spec        object 
 12   sec_spec_1      object 
 13      sec_spec_2   object 
 14   sec_spec_3      object 
 15   sec_spec_4      object 
 16   sec_spec_all    object 
 17      org_nm       object 
 18   org_pac_id      float64
 19   num_org_mem     float64
 20   adr_ln_1        object 
 21   adr_ln_2        object 
 22   ln_2_sprs       object 
 23   cty             object 
 24   st              object 
 25   zip             object 
 26   phn_numbr       float64
 27      hosp_afl

#Above .info() command is not showing the non-null and null values, so I found a possible solution on stackoverflow as shown below


In [ ]:
df.info(verbose=True, show_counts=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2383035 entries, 0 to 2383034
Data columns (total 40 columns):
 #   Column           Non-Null Count    Dtype  
---  ------           --------------    -----  
 0   NPI              2383035 non-null  int64  
 1    Ind_PAC_ID      2383035 non-null  int64  
 2    Ind_enrl_ID     2383035 non-null  object 
 3    lst_nm          2382984 non-null  object 
 4    frst_nm         2383004 non-null  object 
 5    mid_nm          1702897 non-null  object 
 6    suff            38119 non-null    object 
 7    gndr            2383035 non-null  object 
 8    Cred            610734 non-null   object 
 9    Med_sch         2383030 non-null  object 
 10   Grd_yr          2381562 non-null  float64
 11   pri_spec        2383035 non-null  object 
 12   sec_spec_1      325244 non-null   object 
 13      sec_spec_2   34844 non-null    object 
 14   sec_spec_3      3623 non-null     object 
 15   sec_spec_4      615 non-null      object 
 16   sec_spec_all    3

#As shown above, there are a lot of null values in multiple columns 

Shows the first issue of the data, there are a lot of null or empty boxes. 

In [ ]:
#Basic Statistical Data (mean, max, standard deviation, etc)
df.describe()

NPI    Ind_PAC_ID        Grd_yr    org_pac_id   num_org_mem  \
count  2.383035e+06  2.383035e+06  2.381562e+06  2.220212e+06  2.220212e+06   
mean   1.500005e+09  4.993552e+09  2.002790e+03  4.928591e+09  7.115752e+02   
std    2.876755e+08  2.871052e+09  1.243369e+01  2.840208e+09  1.172627e+03   
min    1.003000e+09  4.210003e+07  1.947000e+03  4.210009e+07  2.000000e+00   
25%    1.255309e+09  2.466870e+09  1.994000e+03  2.567460e+09  3.500000e+01   
50%    1.497997e+09  4.981957e+09  2.005000e+03  4.789589e+09  1.990000e+02   
75%    1.750033e+09  7.416947e+09  2.013000e+03  7.416044e+09  9.630000e+02   
max    1.993000e+09  9.931600e+09  2.024000e+03  9.931600e+09  7.567000e+03   

          phn_numbr      hosp_afl_1     hosp_afl_5  
count  2.010498e+06    1.653573e+06  151736.000000  
mean   2.198015e+12    2.641841e+05  287581.218472  
std    1.050912e+14    1.521228e+05  157954.095316  
min    2.004600e+09    1.000100e+04   10001.000000  
25%    3.865031e+09    1.300070e+05  150091.000000  
50%    6.095976e+09    2.500250e+05  281305.000000  
75%    8.027707e+09    3.900490e+05  391308.000000  
max    9.857277e+15    6.703190e+05  670312.000000

In [ ]:
#This shows the amount of null values 
df.isnull().sum()

NPI                      0
 Ind_PAC_ID              0
 Ind_enrl_ID             0
 lst_nm                 51
 frst_nm                31
 mid_nm             680138
 suff              2344916
 gndr                    0
 Cred              1772301
 Med_sch                 5
 Grd_yr               1473
 pri_spec                0
 sec_spec_1        2057791
    sec_spec_2     2348191
 sec_spec_3        2379412
 sec_spec_4        2382420
 sec_spec_all      2057791
    org_nm          162827
 org_pac_id         162823
 num_org_mem        162823
 adr_ln_1                0
 adr_ln_2          1545686
 ln_2_sprs         2232030
 cty                     0
 st                      0
 zip                     0
 phn_numbr          372537
    hosp_afl_1      729462
 hosp_afl_lbn_1     771999
 hosp_afl_2        1578897
 hosp_afl_lbn_2    1591358
 hosp_afl_3        1966949
 hosp_afl_lbn_3    1970488
 hosp_afl_4        2144248
 hosp_afl_lbn_4    2145833
 hosp_afl_5        2231299
 hosp_afl_lbn_5    2232480
 

IMPORTANT SECTIONS WITH NULL (in my opinion)
frst_nm
lst_nm
cred



#Now we can replace or delete all the null values with the mean median mode or using regression. 

(I havent done so as of yet)

In [ ]:
#Checks if our data has any duplicates
df.duplicated(subset = ['NPI'])


0          False
1          False
2          False
3          False
4          False
           ...  
2383030     True
2383031     True
2383032     True
2383033     True
2383034     True
Length: 2383035, dtype: bool

#Our data has no duplicates

In [ ]:
#Checks if all the data for important columns are unique.
df.nunique()

NPI                1212968
 Ind_PAC_ID        1212970
 Ind_enrl_ID       1269278
 lst_nm             274185
 frst_nm             83175
 mid_nm              34643
 suff                   11
 gndr                    2
 Cred                   21
 Med_sch               463
 Grd_yr                 77
 pri_spec               86
 sec_spec_1             70
    sec_spec_2          67
 sec_spec_3             58
 sec_spec_4             44
 sec_spec_all         1553
    org_nm           76036
 org_pac_id          76638
 num_org_mem           681
 adr_ln_1           244741
 adr_ln_2            32171
 ln_2_sprs               1
 cty                 11035
 st                     56
 zip                257681
 phn_numbr          225552
    hosp_afl_1        4819
 hosp_afl_lbn_1       4443
 hosp_afl_2           4758
 hosp_afl_lbn_2       4395
 hosp_afl_3           4625
 hosp_afl_lbn_3       4311
 hosp_afl_4           4406
 hosp_afl_lbn_4       4168
 hosp_afl_5           4142
 hosp_afl_lbn_5       3944
 

#I think that the amount of doctors in the data is 1212968, because thats the total of unique NPI numbers. 

In [ ]:
pd.options.display.max_rows = 100 #changing the maximum amount of rows that are allowed to be displayed from the dataset

print(df.head(100))

#Chose to look at the first 100 to see if I can notice any patterns in the data

           NPI   Ind_PAC_ID      Ind_enrl_ID              lst_nm      frst_nm  \
0   1215155429   6507031499  I20111205000151              WEYMAN        TERRY   
1   1215377635   9133419120  I20200609000011              DUPONT        BRETT   
2   1215151303   4587852165  I20101227000044              MORGAN      BARBARA   
3   1215148507    749332930  I20120404001029            BRANTLEY        ELISE   
4   1215300983   9638476138  I20160329001230             HUNTLEY       TAYLOR   
5   1215197363   5193890473  I20080821000061  CARRERAS RODRIGUEZ         JOHN   
6   1215285135   7214156561  I20140918000582             BLANSON         ERIC   
7   1215203914   4688837818  I20151105000565              LISTER      MICHAEL   
8   1215158522   9234284365  I20090826000098             DEBRUIN        LAURA   
9   1215148200   3375618614  I20080820000611           SHANMUGAM       GANESH   
10  1215306683   8224303789  I20171009003379              JENSEN        JACOB   
11  1215235635   3577503853 

#The middle names of the doctors are either null, the full middle name, or just the initial 

# The 'Ind_PAC_ID' column has varying sizes of the ID numbers, its not consistent as compared to the other ID columns

#Some people have two names in the last name column instead of just one

#The 'cred' column (credentials) has a lot of null values, and it is a pretty important column

#As shown here and in the 'df.isnull().sum()' done above a lot of the 'hospital_lbn' have null values and it may be better if those columns are removed from the data. 

#the assgn and grp_assgn columns show whether the doctor will 'm' for maybe or 'y' for yes will accept payment through a Medicare approved amount. The 'm' doctors can be removed to reduce the pool of doctors to choose from 

###Rather small issue, but there is a typo in the data but thought I should mention it. 


__________________________________________________________________________________________________
#Questions to be answered 

#Where will the experiment be taking place? We can narrow down doctors using a location bias

#Is there an amount of years in practice required? We can narrow down doctors using their graduation year 

#For the doctors with 'other' as the name of their medical school, is the name of the school even necessary and could we remove the medical school column entirely or remove the ones with 'other' as their medical school name. 

#How much is the importance of the 'secondary specialty' columns as they have some of the most null values. 





---------------------------------------------------------------------------------------------

#Suggestions

### Remove the 'suff' column
### Remove the doctors that have a value in 'ln_2_sprs' shows that the address given may be incomplete

### Possibly remove some of the 'hosp_afl_lbn' columns and the 'hosp_afl' columns

